### ISSUE TITLE
Bug: route not recognized in schemas for specific types

### ISSUE THREAD

**Issue by ccrvlh** [litestar-org/litestar#2700]
### Description

Currently some parameters types don't show in the docs. It seems `int` works fine, while `str` and `uuid` don't.
This is silent, no logs are shown, running `debug` mode. Same behavior for both `/schema/elements` and `/schema/swagger`.

### MCVE

```python
(Uncomment one at a time)


class BugRoutes(Controller):
    tags = ["Bugs"]
    path = "/bugs"
    dependencies = dict(context=Provide(route_context))

    @routes.get()
    def get_bugs(self, context: AppContext) -> Response:
        return Response({})

    @routes.post()
    def create_bug(self, context: AppContext, data: Any) -> Response:
        return Response({})

    # This works
    # @routes.get("/{param:int}")
    # def get_bug(self, context: AppContext, param: int) -> Response:
    #    return Response({})

    # This doesn't work (not showing on docs)
    # @routes.get("/{param:str}")
    # def get_bug_str(self, context: AppContext, param: str) -> Response:
    #    return Response({})

    # This doesn't work (not showing on docs)
    # @routes.get("/{param:uuid}")
    # def get_bug_uuid(self, context: AppContext, param: UUID) -> Response:
    #    return Response({})

    @routes.patch("/{param:int}")
    def update_bug(self, context: AppContext, param: int) -> Response:
        return Response({})

    @routes.delete("/{param:int}")
    def delete_bug(self, context: AppContext, param: int) -> None:
        return Response({})
```

### Litestar Version

2.3.2

### Platform

- [X] Mac
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify in the description above)

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh Litestar dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/2700">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/2700/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/litestar-org/litestar/issues/2700/pledge.svg">
</picture>
</a>

**Comment by ccrvlh** [comment]
The behavior is weird, if I have one route with `{param:int}` then all the routes using `{param:uuid}` stop showing and only the `{param:int}` ones show. However if all my routes are `{param:uuid}` then all of them appear.

This happened during development (it's not a common scenario in my case to have routes with multiple param types), so it might be a case where converting the route definition to schema is not "atomic" and one route definition affects the other? If that was the case, there is some sort of implicit priority rule when building the schema, which is also unexpected.

**Comment by Alc-Alc** [comment]
It does appear to be the case where certain routes are not shown in the schema but are indeed accessible (can serve the request to their path).

Posting a slightly modified, runnable version of their code that demonstrates this

```python
from uuid import UUID
from litestar import Controller, Litestar, get, post, patch, delete

class BugRoutes(Controller):
    path = "/bugs"

    @get()
    async def get_bugs(self) -> None:
        return None

    @post()
    async def create_bug(self, data: dict[str, str]) -> None:
        return None

    # This works
    # @get("/{param:int}")
    # async def get_bug(self, param: int) -> None:
    #    return None

    # This doesn't work (not showing on docs)
    # @get("/{param:str}")
    # async def get_bug_str(self, param: str) -> None:
    #    return None

    # This doesn't work (not showing on docs)
    # @get("/{param:uuid}")
    # async def get_bug_uuid(self, param: UUID) -> None:
    #    return None

    @patch("/{param:int}")
    async def update_bug(self, param: int) -> None:
        return None

    @delete("/{param:int}")
    async def delete_bug(self, param: int) -> None:
        return None

app = Litestar([BugRoutes], debug=True)
```

**Comment by provinzkraut** [comment]
@Alc-Alc did you make any progress here?

**Comment by Alc-Alc** [comment]
> @Alc-Alc did you make any progress here?

sorry, missed this notification. Sadly no progress since then. I will try to give it a second look today

**Comment by provinzkraut** [comment]
@Alc-Alc this seems to be related to #2967?

**Comment by Alc-Alc** [comment]
> @Alc-Alc this seems to be related to #2967?

I just ran this on that branch where I fixed that and this is still an issue

**Comment by provinzkraut** [comment]
@Alc-Alc @ccrvlh I can currently not reproduce this behaviour, or I do not understand what exactly the bug is here. Running the examples provided and looking at the generated schema and UI, everything seems to be working as expected. Can you confirm that this bug still exists?

**Comment by provinzkraut** [comment]
Okay, I've managed to reproduce with this:

```python
from litestar import get, post, Litestar


@get("/{param:int}")
async def get_handler() -> None:
    pass


@post("/{param:int}")
async def post_handler() -> None:
    pass


app = Litestar([get_handler, post_handler])
paths = app.openapi_schema.paths["/{param}"]
assert paths.get
assert paths.patch
```

**Comment by provinzkraut** [comment]
Okay, I've managed to reproduce with this:

```python
from litestar import get, post, Litestar


@get("/{param:int}")
async def get_handler() -> None:
    pass


@post("/{param:int}")
async def post_handler() -> None:
    pass


app = Litestar([get_handler, post_handler])
paths = app.openapi_schema.paths["/{param}"]
assert paths.get
assert paths.patch
```

**Comment by provinzkraut** [comment]
@ccrvlh Sorry this took ages to take care of, but I've got a fix pending in #3293, which should be available in the next minor release! :)
